mlrun/api/crud/secrets.py

```diff
@@ -0,0 +1,191 @@
+        allow_internal_secrets: bool = False,
+    ):
+        if secrets.secrets:
+            for secret_key in secrets.secrets.keys():
+                mlrun.utils.helpers.verify_field_regex(
+                    "secret.key", secret_key, mlrun.utils.regex.secret_key
+                )
+                if (
+                    self._is_internal_secret_key(secret_key)
+                    and not allow_internal_secrets
+                ):
+                    raise mlrun.errors.MLRunAccessDeniedError(
+                        f"Not allowed to create/update internal secrets (key starts with "
+                        f"{self.internal_secrets_key_prefix})"
+                    )
+        if secrets.provider == mlrun.api.schemas.SecretProviderName.vault:
+            # Init is idempotent and will do nothing if infra is already in place
+            mlrun.utils.vault.init_project_vault_configuration(project)
+
+            # If no secrets were passed, no need to touch the actual secrets.
+            if secrets.secrets:
+                mlrun.utils.vault.store_vault_project_secrets(project, secrets.secrets)
+        elif secrets.provider == mlrun.api.schemas.SecretProviderName.kubernetes:
+            if mlrun.api.utils.singletons.k8s.get_k8s():
+                mlrun.api.utils.singletons.k8s.get_k8s().store_project_secrets(
+                    project, secrets.secrets
+                )
+            else:
+                raise mlrun.errors.MLRunInternalServerError(
+                    "K8s provider cannot be initialized"
+                )
+        else:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                f"Provider requested is not supported. provider = {secrets.provider}"
+            )
+
+    def delete_secrets(
+        self,
+        project: str,
+        provider: mlrun.api.schemas.SecretProviderName,
+        secrets: typing.Optional[typing.List[str]] = None,
+        allow_internal_secrets: bool = False,
+    ):
+        if not allow_internal_secrets:
+            if secrets:
+                for secret_key in secrets:
+                    if self._is_internal_secret_key(secret_key):
+                        raise mlrun.errors.MLRunAccessDeniedError(
+                            f"Not allowed to delete internal secrets (key starts with "
+                            f"{self.internal_secrets_key_prefix})"
+                        )
+            else:
+                # When secrets are not provided the default behavior will be to delete them all, but if internal secrets
+                # are not allowed, we don't want to delete them, so we list the keys and filter them out
+                secrets_keys_data = self.list_secret_keys(
+                    project, provider, allow_internal_secrets=True
+                )
+                secrets = [
+                    key
+                    for key in secrets_keys_data.secret_keys
+                    if not self._is_internal_secret_key(key)
+                ]
```

Comment:
```suggestion
                secrets = self.list_secret_keys(
                    project, provider, allow_internal_secrets=False
                ).secret_keys
```
You already do this filtering if passing `allow_internal_secrets=False` to the list function 😄